### ISSUE TITLE
Error in Stan code creation when using me()/mi() in non-linear models

### ISSUE THREAD

**Issue by RaoulWolf** [paul-buerkner/brms#810]
*tl;dr: model parsing fails when `me()`/`mi()` are used in non-linear models*

Hi, I'm trying to run a otherwise simple non-linear regression, but the predictor terms are not given individually, but group-wise as mean and standard deviation. To illustrate, I'll modify the example for non-linear modelling with `brms` as given [here](https://cran.r-project.org/web/packages/brms/vignettes/brms_nonlinear.html): 

    library(brms)
    
    b <- c(2, 0.75)
    x0 <- rnorm(100)
    y <- rnorm(100, mean = b[1] * exp(b[2] * x0))
    dat1 <- data.frame(x0, y)
    
    dat2 <- dat1[order(dat1$x0), ]
    rownames(dat2) <- NULL
    
    dat2$gr <- sort(rep(letters[1:5], times = 20))
    
    for (i in dat2$gr) {
      dat2[dat2$gr == i, "x"] <- mean(dat2[dat2$gr == i, "x0"])
      dat2[dat2$gr == i, "sdx"] <- sd(dat2[dat2$gr == i, "x0"])
    }
    
    head(dat2)
    >          x0          y gr         x       sdx
    > 1 -2.443856  0.2633177  a -1.328462 0.4237449
    > 2 -2.058372 -0.1892041  a -1.328462 0.4237449
    > 3 -1.786092 -0.4361064  a -1.328462 0.4237449
    > 4 -1.768300 -0.1509019  a -1.328462 0.4237449
    > 5 -1.518788  1.9475065  a -1.328462 0.4237449
    > 6 -1.482939  0.9069950  a -1.328462 0.4237449

The `x0` predictor is now expressed as group-wise (`gr`) average `x` with standard deviation `sdx`. *For my research, it is unfortunately impossible to obtain any `x0` values, so I'm stuck with `y`, `x`, and `sdx`.* 

I first tried to fit the model from the [example](https://cran.r-project.org/web/packages/brms/vignettes/brms_nonlinear.html), using `x0` as predictor: 

    fit1 <- brm(formula = bf(y ~ b1 * exp(b2 * x0), b1 + b2 ~ 1, nl = TRUE), data = dat2, 
                prior = c(prior(normal(1, 2), nlpar = "b1"), prior(normal(0, 2), nlpar = "b2")))

No problems here. However, when trying to use `x` and `sdx` via the measurement error term (`me()`), the model parsing fails:

    fit2 <- brm(formula = bf(y ~ b1 * exp(b2 * me(x, sdx, gr)), b1 + b2 ~ 1, nl = TRUE), data = dat2, 
                prior = c(prior(normal(1, 2), nlpar = "b1"), prior(normal(0, 2), nlpar = "b2")))
    > SYNTAX ERROR, MESSAGE(S) FROM PARSER:
    > Variable "gr" does not exist.
    >  error in 'model2830170930dc_file283035571e7e' at line 35, column 63
    >  -------------------------------------------------
    >     33:   for (n in 1:N) {
    >     34:     // compute non-linear predictor values
    >     35:     mu[n] = nlp_b1[n] * exp(nlp_b2[n] * me(C_1[n] , C_2[n] , gr=gr));
    >                                                                       ^
    >     36:   }
    >   -------------------------------------------------
    > 
    > Error in stanc(model_code = paste(program, collapse = "\n"), model_name = model_cppname,  : 
    >  failed to parse Stan model 'file283035571e7e' due to the above error.

Removing the `gr` term from the equation (as indicated by the error message) also failed. To make sure that `me()` works , I checked the following dummy model:

    fit3 <- brm(formula = bf(y~ me(x, sdx, gr)), data = dat2)

This model parsed without trouble (but unsurprisingly a convergence warning messages). As the `me()` function only supports a `gaussian` distribution, I also tried to setup the model as bivariate, using a combination of `me()` and `mi()`, and a dummy variable `xnew`:

    dat2$xnew <- NA_real_
    
    fit4 <- brm(formula = mvbf(bf(y ~ b1 * exp(b2 * mi(xnew)), b1 + b2 ~ 1, nl = TRUE), 
                               bf(xnew | mi() ~ me(x, sdx, gr = gr))),
                data = dat2, 
                prior = c(prior(normal(1, 2), nlpar = "b1", resp = "y"), prior(normal(0, 2), nlpar = "b2", resp = "y")))
    > SYNTAX ERROR, MESSAGE(S) FROM PARSER:
    > No matches for: 
    > 
    >   mi(real)
    >  
    > Function mi not found.
    > error in 'model19081f086b32_file19081411d74' at line 66, column 58
    >   -------------------------------------------------
    >     64:   for (n in 1:N_y) {
    >     65:     // compute non-linear predictor values
    >     66:     mu_y[n] = nlp_y_b1[n] * exp(nlp_y_b2[n] * mi(C_y_1[n]));
    >                                                                  ^
    >     67:   }
    >   -------------------------------------------------
    > 
    > Error in stanc(model_code = paste(program, collapse = "\n"), model_name = model_cppname,  : 
    >   failed to parse Stan model 'file19081411d74' due to the above error.

The model parsing fails again; which leaves two possible explanaitons:

* The `brms`-Stan parser mishandles `me()` and `mi()` in the context of non-linear models, or
* It is not possible to use `me()` or `mi()` in the context of non-linear models?

Thank you in advance!

**Comment by paul-buerkner** [comment]
You are using these terms wrong. me() and mi() should be used in linear formulas (of potentially non-linear parameters). You can read more the brms syntax structure in the brms_multilevel vignette or in one of my new papers: https://arxiv.org/abs/1905.09501
